### PR TITLE
PYIC-7109 local F2F

### PR DESCRIPTION
**Pull request by Joe-Edwards-GDS**
### What changed

Set up async credential polling from the queue stub, use it for F2F journeys.

Updated readmes etc. to match.

### Why did it change

So we can test F2F journeys locally

### Issue tracking

- [PYIC-7109](https://govukverify.atlassian.net/browse/PYIC-7109)

[PYIC-7109]: https://govukverify.atlassian.net/browse/PYIC-7109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: api-tests/src/config/config.ts [hunk 1]
```diff
@@ -2,13 +2,12 @@ import dotenv from "dotenv";
...
 
-if (!CORE_ENV) {
-  dotenv.config();
-} else {
+if (CORE_ENV) {
   dotenv.config({
     path: `.env.${CORE_ENV}`,
   });
 }
+dotenv.config();
```
Comment (Joe-Edwards-GDS) [inline api-tests/src/config/config.ts:10]: Config is 'first wins' (by default) - this means that we'll have the precedence:
1. Environment variables
2. `.env.<name>`
3. `.env`

File: api-tests/cucumber.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ export default {
     "json:reports/api-tests-cucumber-report.json",
   ],
   publish: false,
-  retry: 1,
+  retry: 0,
```
Comment (Joe-Edwards-GDS) [inline api-tests/cucumber.js:8]: Wait until we see evidence of flakiness before increasing this...

File: api-tests/README.md [hunk 3]
```diff
@@ -16,42 +16,41 @@
...
-Run using the main `.env` file by using `npm test`.
-
-Additional arguments can be passed to cucumber, e.g. to include specific tags: `npm test -- --tags '@Build'`
-
-You can also run against locally by using the presets:
+There are 3 presets that can be used:
 
+- `npm run test:build` - runs against the deployed build environment
 - `npm run test:local` - runs against an already-running local instance of core-back
 - `npm run test:ci` - starts a new local instance of core-back and tests against it
```
Comment (Joe-Edwards-GDS) [inline api-tests/README.md:23]: (this will require people to change their config, sorry :( )

Comment (Joe-Edwards-GDS) [reply api-tests/README.md:23]: Alternatively we could ask people to set the env var on their system, but I always find that a bit unsatisfactory

File: api-tests/README.md [hunk 4]
```diff
@@ -16,42 +16,41 @@
...
+#### Other environments
 
-For convenience, it is also possible to keep multiple `.env.<name>` files for different environments,
-which are selected by setting `CORE_ENV=<name>` when running the tests.
+It is also possible to set up other `.env.<name>` files,
+which can be selected by setting `CORE_ENV=<name>` when running the tests.
 
-There is a `.env.local` file containing config for a local instance of core-back,
-which is useful for running tests locally or in the pre-merge tests.
+As an example, to run against a deployment in the dev01 account you could create a `.env.dev02` file something like:
```
Comment (thebauSoftwire) [inline api-tests/README.md]: should this be `.env.dev01` instead?

File: api-tests/features/p2-f2f-journey.feature [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    When I submit a 'next' event
+    Then I get a 'claimedIdentity' CRI response
+    When I submit 'kenneth-current' details to the CRI stub
+    Then I get an 'address' CRI response
+    When I submit 'kenneth-current' details to the CRI stub
+    Then I get a 'fraud' CRI response
+    When I submit 'kenneth-score-2' details to the CRI stub
+    Then I get a 'f2f' CRI response
+    When I submit 'kenneth-driving-permit-valid' details to the CRI stub
+    Then I get a 'page-face-to-face-handoff' page response
```
Comment (thebauSoftwire) [inline api-tests/features/p2-f2f-journey.feature:19]: We could use [Background](https://cucumber.io/docs/gherkin/reference/#background) to group these duplicated steps

File: api-tests/features/p2-f2f-journey.feature [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    Then I get a 'page-ipv-identity-postoffice-start' page response
+    When I submit a 'next' event
+    Then I get a 'claimedIdentity' CRI response
+    When I submit 'kenneth-current' details to the CRI stub
+    Then I get an 'address' CRI response
+    When I submit 'kenneth-current' details to the CRI stub
+    Then I get a 'fraud' CRI response
+    When I submit 'kenneth-score-2' details to the CRI stub
+    Then I get a 'f2f' CRI response
+    When I submit 'kenneth-driving-permit-valid' details to the async CRI stub
```
Comment (thebauSoftwire) [inline api-tests/features/p2-f2f-journey.feature]: Should there also be some cases for passport?

Comment (thebauSoftwire) [reply api-tests/features/p2-f2f-journey.feature]: We can use [scenario outline](https://cucumber.io/docs/gherkin/reference/#scenario-outline) to reduce duplication for these

File: api-tests/.env.local [hunk 7]
```diff
@@ -7,3 +7,6 @@ CORE_BACK_PUBLIC_ENCRYPTION_KEY='{"kty": "RSA","n": "0465qJwo8nCkC2tvV4niuWF6IM6
 ORCHESTRATOR_REDIRECT_URL="http://localhost:4500/callback"
 JAR_SIGNING_KEY='{"kty":"EC","d":"OXt0P05ZsQcK7eYusgIPsqZdaBCIJiW4imwUtnaAthU","crv":"P-256","x":"E9ZzuOoqcVU4pVB9rpmTzezjyOPRlOmPGJHKi8RSlIM","y":"KlTMZthHZUkYz5AleTQ8jff0TJiS3q2OB9L5Fw4xA04"}' # pragma: allowlist secret
 LOCAL_AUDIT_EVENTS=true
+ASYNC_QUEUE_DELAY=2
+
+# Also requires ASYNC_QUEUE_NAME to be specified in .env or as an environment variable
```
Comment (thebauSoftwire) [inline api-tests/.env.local:12]: (q for my own learning) does this mean that if we're missing a variable in this .env.local file, it'll look for that variable in the .env file?

**Review by thebauSoftwire — COMMENTED** [review]
nice! Just some small comments + a question

File: api-tests/README.md (hunk 4 above)
Comment (Joe-Edwards-GDS) [reply api-tests/README.md]: 🤦

File: api-tests/.env.local (hunk 7 above)
Comment (Joe-Edwards-GDS) [reply api-tests/.env.local:12]: Yeah, exactly - it'll first check proper environment variables, then `.env.local`, then `.env`

File: api-tests/features/p2-f2f-journey.feature (hunk 5 above)
Comment (Joe-Edwards-GDS) [reply api-tests/features/p2-f2f-journey.feature:19]: Decided not to use a background - we could duplicate the steps up to F2F (CIC/Address/Fraud) but it doesn't feel like a logical 'background' to me. (IMO background should be some common setup, not just a coincidentally identical part of the journey).

It's less bad now I'm using the scenario outline anyway :)

File: api-tests/package.json [hunk 8]
```diff
@@ -5,6 +5,7 @@
   "type": "module",
   "scripts": {
     "test": "cucumber-js",
+    "test:build": "CORE_ENV=build cucumber-js --tags '@Build'",
```
Comment (thebauSoftwire) [inline api-tests/package.json:8]: nice!

**Review by thebauSoftwire — APPROVED** [review]

**Pull request merged**